Design document

kurczynski opened the pull request:
Initial service worker RFD proposal.

kurczynski committed:
Added challenge requirements doc

kurczynski committed:
Initial RFD draft

kurczynski committed:
Added proposed proto files

This helps keep the service's data model consistent between the RFD and
application.

GavinFrazar commented on api/proto/worker/worker.proto:
```diff
@@ -0,0 +1,18 @@
+service Worker {
+  // Gets details of jobs owned by the specified user
+  rpc Query(worker.User) returns (worker.Jobs) {}
+}
```
fetching/listing the jobs owned by a user is out of scope, you can get rid of this.
The API only needs to allow a user to start, stop, stream output, and get the status of an individual job.

kurczynski replied:
Excellent, removed.

GavinFrazar commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,65 @@
+  repeated string args = 2; // Command arguments
+}
+
+enum Statuses {
```
nitpick: singular form makes more sense for an enum, e.g. "running" is a kind of "Status" not a kind of "Statuses"
```suggestion
enum Status{
```

kurczynski replied:
I went back and forth with that a couple of times :smile: Fixed.

GavinFrazar commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,65 @@
+
+message OutputResponse {
+  string stdout = 1;
+  string stderr = 2;
```
what if the user starts a job like `$ cat /dev/random`?

kurczynski replied:
Good catch, updated to use `bytes` instead.

GavinFrazar commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,65 @@
+
+message StartRequest {
+  job.Command command = 1;
+  job.Resources resource_limits = 2;
```
~~I think user-specified resource limits is out of scope - you can hardcode some set of resource limits instead of letting users choose for each job.~~

sclevine replied:
@GavinFrazar Not out of scope for this challenge:

> Add resource control for CPU, Memory and Disk IO **per job** using cgroups.

GavinFrazar commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,136 @@
+  in [Go 1.17](https://go.dev/doc/go1.17#crypto/tls))
+* X.509 public/private key pairs will be generated using RSA-4096
+
+For authorization, the service will use simple tokens to ensure that jobs can only be accessed and modified by the user
```
How are these tokens created?

GavinFrazar replied:
Also, what if another user steals this token? It sounds like the system will reject them anyway since they are not the user who created the job right? Then why have a token at all?

kurczynski replied:
For this challenge, I imagined simply providing any token when a job is created. That token must then be used to stop or query that job. This is definitely useless for real-world use, I can use JSON web tokens if that would be better for this exercise.

GavinFrazar replied:
The token seems unnecessary even for this toy project though - users can only access a job that they created right? And we can verify their identity with mTLS.

kurczynski replied:
Ahhhh okay, that makes more sense to use the existing mTLS. The part that confused me in the challenge doc was `Use mTLS authentication and verify client certificate` followed by `Use a simple authorization scheme`. I took that as "use mTLS and then do something else for authorization".

I'll update the doc to reflect that.

GavinFrazar replied:
mTLS only establishes identity, i.e. authentication, (as does a secret token), so the authorization scheme is really independent from both of these things.

And I think just enforcing that only a job creator can interact with the job is fine as a simple authorization scheme for the challenge

GavinFrazar commented on docs/0001-teleport-job-worker.md:
````diff
@@ -0,0 +1,136 @@
+For example, calling `Query()` in the API is equivalent to calling:
+
+```shell
+job-worker query some-user-token
````
How will the user know the token?

Could you also provide simple examples for using each client command? i.e. start/stop/query/output

kurczynski replied:
The token is just any they would make up for this challenge, I can change that though if needed per the other [comment thread](https://github.com/kurczynski/teleport-job-worker/pull/1#discussion_r1610801103).

Certainly, I'll add each of the commands.

GavinFrazar replied:
I do think users need some way to refer to a job, but what if two users choose the same "token" to use?

kurczynski replied:
Dropping the token to use just mTLS will solve this.

GavinFrazar commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,136 @@
+* `Start` start a new job and begin execution of the specified command immediately
+* `Stop` stop execution of the specified job immediately
+* `Query` query details about specified job; this function can run on a job of any status
+* `Output` get the full output (stdout and stderr) of any existing job
```
Some more detail here would be nice - how will concurrent output be handled? How will you stream the output from process start? What happens when the job is no longer running?

kurczynski replied:
Sounds good. I updated this with a bit more detail, I can dive deeper if it's useful.

kurczynski committed:
Removed the worker proto spec

This is out of scope for the challenge, only single jobs need to be
managed.

kurczynski committed:
Fixed naming for job enum

sclevine commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,65 @@
+}
+
+message Resources {
+  int32 memory = 1;
```
Would it make sense to give these names that better indicate how they apply controls?

Could you add a comment documenting each field?

kurczynski replied:
Yeah, I think that's a good idea. Done.

sclevine commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,127 @@
+like `/sys/fs/cgroup/job-worker/job-a`. Under this path, the following files would be updated to manage the resources
+for the job:
+
+* `cgroup.procs` contains the PID of the job's process being managed
```
How will you ensure the cgroup applies to this PID for the entirety of its execution?

kurczynski replied:
I'm not sure I understand the question. The job will fork from the service, the job will apply the cgroups, and then execute the process given by the user.

The kernel [docs](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#processes) say:
> When a process forks a child process, the new process is born into the cgroup that the forking process belongs to at the time of the operation. After exit, a process stays associated with the cgroup that it belonged to at the time of exit until it’s reaped

So it should be applied the whole time.

sclevine replied:
>  The job will fork from the service, the job will apply the cgroups, and then execute the process given by the user.

Awesome. Might want to make this a little more clear. It's clear to me now, re-reading:

> Jobs will be forked from the service's process and contain exactly one process that it manages.

But for a while I thought your service was managing the command process directly.

sclevine commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,65 @@
+}
+
+message Resources {
+  int32 memory = 1;
```
Is int32 appropriately sized for this value?

kurczynski replied:
It would probably provide enough memory, but I agree, better to give access to all the memory with `uint64`.

kurczynski committed:
Changed job output response type

Use bytes instead of strings to better handle non-printable output from
commands such as 'cat /dev/random'.

kurczynski committed:
Added more details on the CLI functionality

kurczynski committed:
Updated resource names and descriptions

Makes it easier to understand what the resources control and how they
are applied.

kurczynski committed:
Updated memory resource from int32 to uint64

Since we can only specify up to about 2GB with int32, use uint64 instead
to allow specifying larger memory values.

kurczynski committed:
Added more details to API

kurczynski committed:
Added timestamps and another status to the protobuf

Timestamps about job start and stop times may be useful when interacting
with jobs.

The "READY" status has been added to account for the time between when a
job is created and when command execution has started.

kurczynski committed:
Added library godoc to the RFD

sclevine commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,185 @@
+}
+    Job Contains information to interact with jobs.
+
+func GetJob(id string) *Job
```
This method implies that the library stores a global map of jobs. Is this necessary?

kurczynski replied:
That's what I had in mind since we want to query the jobs. I guess an important question would be, what kind of information do we want returned from a query call?

sclevine replied:
I would avoid global state in the library.

While the API needs to keep a store of jobs, I don't see a similar requirement for the library.

sclevine commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,185 @@
+    exists.
+
+func NewJob(command string, args ...string) *Job
+    NewJob Create a new job to run the specified command.
```
```suggestion
// NewJob Create a new job to run the specified command.
func NewJob(command string, args ...string) *Job
```

sclevine commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,185 @@
+func NewJob(command string, args ...string) *Job
+    NewJob Create a new job to run the specified command.
+
+func (j *Job) Id() string
```
```suggestion
func (j *Job) ID() string
```
https://go.dev/wiki/CodeReviewComments#initialisms

sclevine commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,72 @@
+}
+
+message Resources {
+  uint64 memory_bytes = 1; // Amount of memory in bytes that a job can use
```
```suggestion
  // Amount of memory in bytes that a job can use
  uint64 memory_bytes = 1;
```

sclevine commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,72 @@
+message Command {
+  string name = 1; // Name of the command to execute
+  repeated string args = 2; // Command arguments
+  uint64 start_time = 3; // Time the command started in number of milliseconds since the epoch
```
Note that you can use the google.protobuf types if you wish: https://protobuf.dev/reference/protobuf/google.protobuf/#timestamp

kurczynski replied:
Fantastic! I found that but thought it would be considered an external dependency.

GavinFrazar commented on docs/0001-teleport-job-worker.md:
````diff
@@ -0,0 +1,185 @@
+  ```shell
+  job-worker -token "some-token" start -cpu 10 -memory 1000 -io 400 -cmd "/usr/bin/sleep" -args "60"
+  ```
+  The auto-generated ID for the job will be returned if the job is started successfully. The resource arguments use the
````
can you elaborate on how the ID is generated and why it's generated that way?

kurczynski replied:
Of course. Done.

GavinFrazar commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,185 @@
+* TLS using the modern version 1.3
+* The cipher suite will need no configuration, it will be handled by `crypto/tls` (this was updated
+  in [Go 1.17](https://go.dev/doc/go1.17#crypto/tls))
+* X.509 public/private key pairs will be generated using RSA-4096
```
For this challenge I think it's ok to just pre-generate the key pairs for client and server, but can you provide explanation for how you will generate all the key pairs necessary to interact with the service, like specific commands and settings you will use?

Preferably as a script (you can link to it instead of copying it inline)

kurczynski replied:
Certainly, I'll link the `Makefile` if that works.

kurczynski committed:
Updated process execution section

Clarified how resources will be controlled for jobs.

bernardjkim commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,72 @@
+  uint64 start_time = 3; // Time the job started in number of milliseconds since the epoch
+  uint64 finish_time = 4; // Time the job stopped in number of milliseconds since the epoch
+  job.Command command = 5;
+  string status_info = 6; // Information about the current job status; may be empty
```
What additional info will this provide over `status`?

kurczynski replied:
It would provide more information about a job's status as needed. For example, if a job's state is `failed` the `status_info` field might have some human-readable string like `exceeded memory limit`.

Maybe `status_details` or `status_reason` would be better?

bernardjkim replied:
Gotcha. The name is okay. It could be helpful to add this context to the comments.

bernardjkim commented on api/proto/job/job.proto:
```diff
@@ -0,0 +1,72 @@
+message Command {
+  string name = 1; // Name of the command to execute
+  repeated string args = 2; // Command arguments
+  uint64 start_time = 3; // Time the command started in number of milliseconds since the epoch
```
This is a little confusing because `start_time` initially makes me think the `Command` can be scheduled to start at this time.
 
Might be better to move this from `Command` to `Info` as `command_start_time`. Or maybe track status transitions with timestamps.

kurczynski replied:
I wanted to have two distinct sets of timestamps for the job and the command without just tossing them all into `Info`. I like the idea of tracking status transitions with timestamps because I think it accomplishes this and provides more details. I'll update the proto :+1:

bernardjkim commented on docs/0001-teleport-job-worker.md:
````diff
@@ -0,0 +1,186 @@
+
+* **Start**
+  ```shell
+  job-worker -token "some-token" start -cpu 10 -memory 1000 -io 400 -cmd "/usr/bin/sleep" -args "60"
````
Looks like tokens are no longer going to be used. Might want to remove the `-token` flags from the UX design.

kurczynski committed:
Updated proto formatting and data types

Fixed formatting to follow convention and changed data types for
timestamps from milliseconds to timestamp because it's way easier to
use.

kurczynski committed:
Fixed godoc formatting

kurczynski committed:
Updated design details and added Makefile

The Makefile generates certs and proto files.

kurczynski committed:
Updated how command info is measured

Measure when a job's status changes to provide more detail about the
job.

GavinFrazar commented on Makefile:
```diff
@@ -0,0 +1,41 @@
+		-out $(CERT_PATH)/$(SERVER)-cert.pem \
+		-keyout $(CERT_PATH)/$(SERVER)-key.pem \
+		-addext "subjectAltName = DNS:$(CERT_HOST)" \
+		-subj "/C=US/ST=California/L=Somewhere/O=My Organization/OU=My Unit/CN=$(CERT_HOST)"
```
last thing, but I won't block the approval over this detail:

with this setup the server and client certs are both CA certs

```fish
$ openssl x509 -in client-cert.pem -noout -text | grep CA
                CA:TRUE
```

This won't be secure for mTLS - let's update it so the client and server certs are not CAs

kurczynski replied:
Good call, I've updated the Makefile.

kurczynski committed:
Added more detail to the job proto

kurczynski committed:
Updated cert generation config

When generating certificates, don't generate them as CAs. This would
make them insecure for mTLS.

sclevine commented on docs/0001-teleport-job-worker.md:
````diff
@@ -0,0 +1,185 @@
+
+### Library Godoc
+
+```
````
Looks like the library is missing configuration for resource controls?

kurczynski replied:
~I was planning on using the `Resources` defined in the proto file. Would you like me to add that section to this part of the document?~

Nevermind, based on your [other comment](https://github.com/kurczynski/teleport-job-worker/pull/1#discussion_r1617674569) I'll add a resource definition.

sclevine commented on docs/0001-teleport-job-worker.md:
```diff
@@ -0,0 +1,185 @@
+func (j *Job) Output() *job.OutputRequest
+
+// Query Get information about the job.
+func (j *Job) Query() *job.Info
```
Library should be separate from the API server / not coupled to gRPC.

(Assuming `job` is a generated protobuf package.)

kurczynski replied:
Good thinking, I'll update the library section.

sclevine reviewed (approved):
Okay to move forward to implementation, but note remaining nits.

kurczynski committed:
Decoupled library from API and formatting

Keep as much logic separated as possible from the library and API in
order to minimize their unintended impact on each other.
